### PR TITLE
[C++20][Modules] Fix crash when function and lambda inside loaded from different modules

### PR DESCRIPTION
**Pull request by dmpolukhin**
Summary:
Because AST loading code is lazy and happens in unpredictable order, it is possible that a function and lambda inside the function can be loaded from different modules. As a result, the captured DeclRefExpr won’t match the corresponding VarDecl inside the function. This situation is reflected in the AST as follows:

```
FunctionDecl 0x555564f4aff0 <Conv.h:33:1, line:41:1> line:33:35 imported in ./thrift_cpp2_base.h hidden tryTo 'Expected<Tgt, const char *> ()' inline
|-also in ./folly-conv.h
`-CompoundStmt 0x555564f7cfc8 <col:43, line:41:1>
  |-DeclStmt 0x555564f7ced8 <line:34:3, col:17>
  | `-VarDecl 0x555564f7cef8 <col:3, col:16> col:7 imported in ./thrift_cpp2_base.h hidden referenced result 'Tgt' cinit
  |   `-IntegerLiteral 0x555564f7d080 <col:16> 'int' 0
  |-CallExpr 0x555564f7cea8 <line:39:3, col:76> '<dependent type>'
  | |-UnresolvedLookupExpr 0x555564f7bea0 <col:3, col:19> '<overloaded function type>' lvalue (no ADL) = 'then_' 0x555564f7bef0
  | |-CXXTemporaryObjectExpr 0x555564f7bcb0 <col:25, col:45> 'Expected<bool, int>':'folly::Expected<bool, int>' 'void () noexcept' zeroing
  | `-LambdaExpr 0x555564f7bc88 <col:48, col:75> '(lambda at Conv.h:39:48)'
  |   |-CXXRecordDecl 0x555564f76b88 <col:48> col:48 imported in ./folly-conv.h hidden implicit <undeserialized declarations> class definition
  |   | |-also in ./thrift_cpp2_base.h
  |   | `-DefinitionData lambda empty standard_layout trivially_copyable literal can_const_default_init
  |   |   |-DefaultConstructor defaulted_is_constexpr
  |   |   |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  |   |   |-MoveConstructor exists simple trivial needs_implicit
  |   |   |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
  |   |   |-MoveAssignment
  |   |   `-Destructor simple irrelevant trivial constexpr needs_implicit
  |   `-CompoundStmt 0x555564f7d1a8 <col:58, col:75>
  |     `-ReturnStmt 0x555564f7d198 <col:60, col:67>
  |       `-DeclRefExpr 0x555564f7d0a0 <col:67> 'Tgt' lvalue Var 0x555564f7d0c8 'result' 'Tgt' refers_to_enclosing_variable_or_capture
  `-ReturnStmt 0x555564f7bc78 <line:40:3, col:11>
    `-InitListExpr 0x555564f7bc38 <col:10, col:11> 'void'
```

This diff modifies the AST deserialization process to load lambdas within the canonical function declaration sooner, immediately following the function, ensuring that they are loaded from the same module.

Re-land https://github.com/llvm/llvm-project/pull/104512 Added test case that caused crash due to multiple enclosed lambdas deserialization.

Test Plan: check-clang

### CODE REVIEW

**Comment by ChuanqiXu9** [comment]
Would you like to explain more why this fail previously in more detail?

Also I am thinking if we can make the process more efficiently:
(1) Can we avoid the visitor in the writing process?
(2) Can we delay the loading of lambdas to the load of definitions of the functions?

I immediately expectation is that (1) is possible but (2) may not be not safe.

For (1), my idea is to record/register the information during the writing of LambdaExpr in ASTWriterStmt and then we can write this record after we write all the decls and types. Then we need to read this record eagerly. Maybe we can optimize it further by adding a slot in FunctionDecl to record the offset of such informations. So we can avoid the eager linear reading process.

For (2), whether or not it is safe is, can we we load the lambdas without loading the definition of the functions? If not, we can add a code when getting the definition of the functions like:

```
if (lambdas not loaded)
     getCanonicalDecl()->loadLambdas();
```

Then we can delay the loading of the lambdas to make it more efficient.

**Comment by dmpolukhin** [comment]
With DeclContext visitor collectLambdas takes 0.03% or smaller, sometime I don't even see it in sampling profile.

**Comment by ChuanqiXu9** [comment]
> > Would you like to explain more why this fail previously in more detail?
> 
> Original code in `ASTReader::finishPendingActions` looked like this:
> 
> ```
>     for (auto ID : PendingLambdas)
>       GetDecl(ID);
>     PendingLambdas.clear();
> ```
> 
> The issue here is that the code uses implicit iterators for `PendingLambdas`, but `GetDecl` may insert more elements into `PendingLambdas`, which can invalidate the iterators. In a good case, when there is no vector relocation, the new elements will be skipped. However, in the reproducer, the insertion caused the vector to relocate, resulting in a crash due to reading invalid values from deallocated memory. To address this issue, I have enclosed more than 5 lambdas to cause relocation in the new test cases.
> 
> In the new code, before reading the lambdas, I copy the existing lambdas to a new array. Alternatively, I could use an integer index for iteration and read the size of the vector on each iteration. Both approaches work fine, but I decided that running other pending actions might be better before starting to deserialize new lambdas. I can change it if you think iteration with an integer index is better.

Thanks, it is clear.

> 
> > Also I am thinking if we can make the process more efficiently: (1) Can we avoid the visitor in the writing process? (2) Can we delay the loading of lambdas to the load of definitions of the functions?
> > I immediately expectation is that (1) is possible but (2) may not be not safe.
> > For (1), my idea is to record/register the information during the writing of LambdaExpr in ASTWriterStmt and then we can write this record after we write all the decls and types. Then we need to read this record eagerly. Maybe we can optimize it further by adding a slot in FunctionDecl to record the offset of such informations. So we can avoid the eager linear reading process.
> 
> Yeah, it will complicate things a lot, we visit statements only after serializing FunctionDecl. I run some profiling and it seems that `collectLambdas` inclusively takes only about 0.25% cycles own time is almost 0. I'll try to reduce it even further. I think we can scopes instead of statements - it should be more efficient and we do it for enumerating anonymous definitions and it seems to be fast enough.

it won't be much more complicated. You only need:
(1) Add a new map in ASTWriter. 
(2) In ASTWriterDecl, when you are writing a CXXRecord you can judge if it is lambda and if its enclosing decl context is a first function decl. And if all the conditions are tree, insert the id of function decl and the record id to that map. (I thought we could do this in ASTWriterStmt, but it may be more straight forward in ASTWriterDecl)
(3) In `ASTWriter::WriteDeclAndTypes`, after we set `DoneWritingDeclsAndTypes = true`, there are plenty examples that we did similar things here (write the information recorded during writing). We can add a new logic below
(4) Convert the map to RecordData (a vector), and then we can emit the record.
(5) In the reader side, convert the readed RecordData into a map.
(6) When we reads the corresponding function (or other possible conditions), loads the lambdas from the map.

I think the steps here are clear enough. On the one hand, 0.3% is not too small, we see 0.5% as significant in middle end. On the other hand, the steps are pretty common in Serializations. I think it will be helpful for you to understand the process.

Sorry in a head that I approve the previous solution but now ask for something more complex. But let's try to do things better if possible.

> 
> > For (2), whether or not it is safe is, can we we load the lambdas without loading the definition of the functions? If not, we can add a code when getting the definition of the functions like:
> > ```
> > if (lambdas not loaded)
> >      getCanonicalDecl()->loadLambdas();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then we can delay the loading of the lambdas to make it more efficient.
> 
> It will be too late in my example. Without my change lambdas get loaded during function body deserialization but it was too late because loading template specialization caused choosing canonical record for the lambda from another modules. Function itself is selected from the last loaded modules by lookup (i.e. it happens in reverse order of module loading) but template specialisations get loaded in direct order. I tried to change order of specilization loading to opposite but in real word it not always works because some modules can be loaded for example in case of `-fmodule-file=<module-name>=<path/to/BMI>`.

So in short, in my question, can I treat your answer as, we can load a lambda or choose a lambda as the canonical decl without deserializing the **corresponding** body of the function decl? And if yes, can we try to explore in which situations? It will be helpful to optimize our strategy.

**Comment by dmpolukhin** [comment]
@ChuanqiXu9 thank you for the detailed steps, I uploaded new version. Does it look like something you had in mind?

File: clang/lib/Serialization/ASTReaderDecl.cpp [hunk 1]
```diff
@@ -4351,6 +4351,15 @@ void ASTReader::loadDeclUpdateRecords(PendingUpdateRecord &Record) {
           reader::ASTDeclContextNameLookupTrait(*this, *Update.Mod));
     DC->setHasExternalVisibleStorage(true);
   }
+
+  // Load any pending lambdas for the function.
+  if (auto *FD = dyn_cast<FunctionDecl>(D); FD && FD->isCanonicalDecl()) {
+    if (auto IT = FunctionToLambdasMap.find(ID);
+        IT != FunctionToLambdasMap.end()) {
+      for (auto LID : IT->second)
+        GetDecl(LID);
```
Comment (ChuanqiXu9) [inline clang/lib/Serialization/ASTReaderDecl.cpp:4360]: Then, let's erase `IT`.

File: clang/include/clang/Serialization/ASTBitCodes.h [hunk 2]
```diff
@@ -724,6 +724,9 @@ enum ASTRecordTypes {
 
   /// Record code for vtables to emit.
   VTABLES_TO_EMIT = 70,
+
+  /// Record code for the FunctionDecl to lambdas mapping.
```
Comment (ChuanqiXu9) [inline clang/include/clang/Serialization/ASTBitCodes.h:728]: Let's add some simple comments here to explain why we want to do this.

File: clang/include/clang/Serialization/ASTReader.h [hunk 3]
```diff
@@ -532,6 +532,10 @@ class ASTReader
   /// namespace as if it is not delayed.
   DelayedNamespaceOffsetMapTy DelayedNamespaceOffsetMap;
 
+  /// Mapping from FunctionDecl IDs to the corresponding lambda IDs.
```
Comment (ChuanqiXu9) [inline clang/include/clang/Serialization/ASTReader.h:535]: Same here. Let's explain our intention.

File: clang/include/clang/Serialization/ASTWriter.h [hunk 4]
```diff
@@ -233,6 +233,10 @@ class ASTWriter : public ASTDeserializationListener,
   /// instead of comparing the result of `getDeclID()` or `GetDeclRef()`.
   llvm::SmallPtrSet<const Decl *, 32> PredefinedDecls;
 
+  /// Map that provides the ID of function to the vector of lambdas inside it.
```
Comment (ChuanqiXu9) [inline clang/include/clang/Serialization/ASTWriter.h:236]: Same here. I don't hate some redundancy.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM with nits

**Review by dmpolukhin — COMMENTED** [review]
All comments resolved, also I added TODO that we might want to use on disk hash table to allow lazy deserialization for the new record. I think performance difference should be negligible but I haven't tested this hypothesis.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM then

**Comment by eaeltsin** [comment]
Heads-up - this commit might have introduced compilation non-determinism.

Will try to cook the reproducer soon. Meanwhile, can you please double-check there are no non-deterministic pieces, such as sorting by address, etc.?

**Comment by eaeltsin** [comment]
A comment on another non-determinism issue that might be related - https://github.com/llvm/llvm-project/issues/106855#issuecomment-2322927480

**Comment by dmpolukhin** [comment]
@eaeltsin do you know which artifacts become non-deterministic? Is it produced object file or serialized AST?

**Comment by eaeltsin** [comment]
We see non-determinism in pcm files when compiling some stl stuff.

Bisection pointed to this commit, but there is still some chance this is because of our internal changes though.

**Comment by dmpolukhin** [comment]
I think it is due to this change :( I'm iterating over `Decl*` in `FunctionToLambdasMap`. I'm fixing it by replacing the pointers with LocalDeclID that should be stable.

**Comment by eaeltsin** [comment]
Perfect, thanks!

File: clang/lib/Serialization/ASTWriter.cpp [hunk 5]
```diff
@@ -5706,6 +5707,27 @@ void ASTWriter::WriteDeclAndTypes(ASTContext &Context) {
     Stream.EmitRecord(DELAYED_NAMESPACE_LEXICAL_VISIBLE_RECORD,
                       DelayedNamespaceRecord);
 
+  if (!FunctionToLambdasMap.empty()) {
+    // TODO: on disk hash table for function to lambda mapping might be more
+    // efficent becuase it allows lazy deserialization.
+    RecordData FunctionToLambdasMapRecord;
+    for (const auto &Pair : FunctionToLambdasMap) {
```
Comment (rnk) [inline clang/lib/Serialization/ASTWriter.cpp:5714]: This looks like the non-deterministic for loop, this is the DenseMap iteration that decides what to write to disk. I'll just send a patch to make this a MapVector.

Comment (rnk) [reply clang/lib/Serialization/ASTWriter.cpp:5714]: Nevermind, I see you have a better fix.

**Comment by aeubanks** [comment]
this has also caused the following issue to pop up in our modules build

```
error: variable 'param' cannot be implicitly captured in a lambda with no capture-default specified
```

hopefully the error provides some guidance for where to look, but I can try to get a reduced repro

**Comment by dmpolukhin** [comment]
> this has also caused the following issue to pop up in our modules build
> 
> ```
> error: variable 'param' cannot be implicitly captured in a lambda with no capture-default specified
> ```
> 
> hopefully the error provides some guidance for where to look, but I can try to get a reduced repro

Please create a reproducer without it is not possible to debug and fix. On a build without asserts my original example gives the same error. It means that there are cases when some lambda is not deserialized early enough.

**Comment by dmpolukhin** [comment]
> still trying to reduce the reproducer...

Please create a reproducer, the stack trace doesn't have information to understand what is causing the issue.

**Comment by ilya-biryukov** [comment]
Here is the reproducer that crashes at head.
[lambdas.tgz](https://github.com/user-attachments/files/17329901/lambdas.tgz)

**Comment by dmpolukhin** [comment]
> Here is the reproducer that crashes at head. [lambdas.tgz](https://github.com/user-attachments/files/17329901/lambdas.tgz)

I reproduced this issue on my machine and confirm that it is indeed due to my changes. https://github.com/llvm/llvm-project/pull/111992 has fix that seems to fix the reproducer and doesn't introduce new issue as far as I know.
@ilya-biryukov could you please try it on your full test case?

**Comment by ilya-biryukov** [comment]
> > Here is the reproducer that crashes at head. [lambdas.tgz](https://github.com/user-attachments/files/17329901/lambdas.tgz)
> 
> I reproduced this issue on my machine and confirm that it is indeed due to my changes. #111992 has fix that seems to fix the reproducer and doesn't introduce new issue as far as I know. @ilya-biryukov could you please try it on your full test case?

I've checked #111992 and it does fix the problem, so let's land it?

**Comment by dmpolukhin** [comment]
> I've checked #111992 and it does fix the problem, so let's land it?

Yes, I would like to create a test case to don't regress this feature in future. I need to reduce libc++ `functional` header to something smaller.

**Comment by ilya-biryukov** [comment]
> > I've checked #111992 and it does fix the problem, so let's land it?
> 
> Yes, I would like to create a test case to don't regress this feature in future. I need to reduce libc++ `functional` header to something smaller.

Right, definitely +1 to adding a test case. I decided to hold off that comment until the PR moves away from the draft state, because I thought that's something that's on your mind anyway.
